piccolo_api/crud/endpoints.py

```diff
@@ -478,19 +506,101 @@ async def _get_all(
+            # for preventing IndexError in piccolo_admin filtering
+            next_cursor = ""
+
+        if cursor:
+            # query parameter cursor
+            next_cursor = self.decode_cursor(cursor)
+            # querying by query parameter order
+            if order_by and order_by.ascending:
+                # use __previous=yes param for getting previous page results
+                # from cursor. For next page we don't pass anything to the query
+                # parameter to get correct next page despite to order param
+                if "__previous" in params:
+                    query = (
+                        query.where(self.table.id < int(next_cursor))
+                        .order_by(self.table.id, ascending=False)
+                        .limit(page_size)
+                    )
+                else:
+                    query = query.where(
+                        self.table.id >= int(next_cursor)
+                    ).limit(page_size + 1)
+            else:
+                if "__previous" in params:
+                    query = (
+                        query.where(self.table.id > int(next_cursor))
+                        .order_by(self.table.id, ascending=True)
+                        .limit(page_size)
+                    )
+                else:
+                    query = query.where(
+                        self.table.id <= int(next_cursor)
+                    ).limit(page_size + 1)
+        rows = await query.run()
+        # if no more further results set next_cursor to first value
+        # from latest rows to get correct previous page (if we provide
+        # __previous=yes) rows and handle edge case if page size gt or gte
+        if len(rows) <= page_size or page_size > len(rows):
+            next_cursor = self.encode_cursor(str(rows[0]["id"]))
+        else:
+            next_cursor = self.encode_cursor(str(rows[-1]["id"]))
+
         query = query.limit(page_size)
         page = split_params.page
         if page > 1:
             offset = page_size * (page - 1)
             query = query.offset(offset).limit(page_size)
 
-        rows = await query.run()
+        # if we use cursor and previous param reverse order of
+        # rows if we go to previous page at correct order opposite
+        # from order param. If we use page param use normal rows
+        # for LimitOffset pagination
+        if "__previous" in params:
+            rows = list(reversed(rows))
+            next_cursor = self.encode_cursor(str(rows[0]["id"]))
+        else:
+            rows = await query.run()
+
         # We need to serialise it ourselves, in case there are datetime
         # fields.
+        cursor_model = Cursor(next_cursor=next_cursor).json()
         json = self.pydantic_model_plural(include_readable=include_readable)(
             rows=rows
         ).json()
-        return CustomJSONResponse(json)
+        return CustomJSONResponse(f"{json[:-1]}, {cursor_model[1:]}")
```

Comment:
What do you think about returning the cursor as a header, rather than in the body? It might simplify this bit of code?

I'm not saying it's right - just interested in your thoughts.

Reply:
@dantownsend Whatever you want. I think it’s simpler in the body but you know better :)
```
Example of Vue frontend

getPosts() {
      const path = "/posts/?__page_size=5&__order=id&__cursor=" + this.cursor + "&__readable=true";
      axios
        .get(path)
        .then(res => {
          this.posts = res.data.rows;
          this.cursor = res.data.next_cursor;  // use cursor from request body like any other variable
        })
        .catch(error => {
          console.error(error);
        });
    },
```
If you want to return cursor from header, can you implement that?

Reply:
Should just be `res.headers.Next-Cursor`, assuming the header is called `Next-Cursor`.

Reply:
Did you mean something like this:
```
# set cursor in header
headers: t.Dict[str, str] = {}
headers["cursor"] = next_cursor

# We need to serialise it ourselves, in case there are datetime
# fields.
json = self.pydantic_model_plural(include_readable=include_readable)(
    rows=rows
).json()
return CustomJSONResponse(json, headers=headers)
```
After that we can access cursor with `res.headers['cursor']` (not `res.headers.cursor`) but we first need to add `expose_headers=["cursor"]` in `CORSMiddleware`. If we don't do this the axios doesn't return the cursor from headers. If we use a cursor from the request body we don’t have to do any of that.

Reply:
Yeah that's a good point - I didn't consider CORS.

OK, so let's keep it in the response body.

We could potentially tidy it up a bit by modifying `pydantic_model_plural`, so it contains the cursor attribute:

```python
    def pydantic_model_plural(self, include_readable=False):
        """
        This is for when we want to serialise many copies of the model.
        """
        base_model = create_pydantic_model(
            self.table,
            include_default_columns=True,
            include_readable=include_readable,
            model_name=f"{self.table.__name__}Item",
        )
        return pydantic.create_model(
            str(self.table.__name__) + "Plural",
            __config__=Config,
            rows=(t.List[base_model], None),
            cursor=(t.Optional[str], None)
        )
```
That way we wouldn't need the separate `Cursor` model, and to combine the JSON strings. Instead it would just be:

```python
return CustomJSONResponse(
    self.pydantic_model_plural(include_readable=include_readable)(
        rows=rows,
        cursor=cursor
    ).json()
)
```
What do you think?

Reply:
This is great. I just change one line and everything works great. Thanks.

```
return CustomJSONResponse(
    self.pydantic_model_plural(include_readable=include_readable)(
        rows=rows,
        cursor=next_cursor // I changed this line 
    ).json()
)
```
Now we don't need `Cursor` model and this weird line `return CustomJSONResponse(f"{json[:-1]}, {cursor_model[1:]}")`